In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# read data
data = pd.read_csv('voice.csv', encoding='latin-1')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3168 entries, 0 to 3167
Data columns (total 21 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   meanfreq  3168 non-null   float64
 1   sd        3168 non-null   float64
 2   median    3168 non-null   float64
 3   Q25       3168 non-null   float64
 4   Q75       3168 non-null   float64
 5   IQR       3168 non-null   float64
 6   skew      3168 non-null   float64
 7   kurt      3168 non-null   float64
 8   sp.ent    3168 non-null   float64
 9   sfm       3168 non-null   float64
 10  mode      3168 non-null   float64
 11  centroid  3168 non-null   float64
 12  meanfun   3168 non-null   float64
 13  minfun    3168 non-null   float64
 14  maxfun    3168 non-null   float64
 15  meandom   3168 non-null   float64
 16  mindom    3168 non-null   float64
 17  maxdom    3168 non-null   float64
 18  dfrange   3168 non-null   float64
 19  modindx   3168 non-null   float64
 20  label     3168 non-null   obje

In [5]:
# encode kolom label
from sklearn.preprocessing import StandardScaler, LabelEncoder
le = LabelEncoder()
data['label'] = le.fit_transform(data['label'])
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000      1  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632      1  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512      1  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119      1  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274      1  

[5 rows x 21 columns]

### Tugas 1, Buatlah model klasifikasi dengan menggunakan SVM untuk data suara

In [9]:
# Buatlah model klasifikasi dengan menggunakan SVM untuk data suara
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix

# memisahkan data independent dan dependent
X = data.drop(['label'], axis=1)
y = data['label']

# normalisasi data
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)

# split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# train model
model = SVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[309   7]
 [  6 312]]
              precision    recall  f1-score   support

           0       0.98      0.98      0.98       316
           1       0.98      0.98      0.98       318

    accuracy                           0.98       634
   macro avg       0.98      0.98      0.98       634
weighted avg       0.98      0.98      0.98       634


### Tugas 2, Buatlah model klasfikasi Multinomial Naive Bayes untuk data suara

In [15]:
# Buatlah model klasfikasi Multinomial Naive Bayes untuk data spam.csv
# Fitur CountVectorizer dengan mengaktifkan stop_words
# Evaluasi hasilnya

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

# read data
data = pd.read_csv('spam.csv', encoding='latin-1')
data.info()

# encode kolom label
le = LabelEncoder()
data['v1'] = le.fit_transform(data['v1'])
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


v1                                                 v2 Unnamed: 2  \
0   0  Go until jurong point, crazy.. Available only ...        NaN   
1   0                      Ok lar... Joking wif u oni...        NaN   
2   1  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   0  U dun say so early hor... U c already then say...        NaN   
4   0  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [17]:
# mengubah nama kolom v1 menjadi labels dan v2 menjadi sms
data = data.rename(columns={'v1': 'labels', 'v2': 'sms'})

# memisahkan data independent dan dependent
X = data['sms']
y = data['labels']

# normalisasi data
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(X)

# split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# train model
model = MultinomialNB()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[945  11]
 [  9 150]]
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       956
           1       0.93      0.94      0.94       159

    accuracy                           0.98      1115
   macro avg       0.96      0.97      0.96      1115
weighted avg       0.98      0.98      0.98      1115


### Tugas 3, Buatlah model klasfikasi Multinomial Naive Bayes 

In [18]:
# Buatlah model klasfikasi Multinomial Naive Bayes untuk data spam.csv
# Fitur TfidfVectorizer dengan mengaktifkan stop_words
# Evaluasi hasilnya

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB

# read data
data = pd.read_csv('spam.csv', encoding='latin-1')
data.info()

# encode kolom label
le = LabelEncoder()
data['v1'] = le.fit_transform(data['v1'])
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


v1                                                 v2 Unnamed: 2  \
0   0  Go until jurong point, crazy.. Available only ...        NaN   
1   0                      Ok lar... Joking wif u oni...        NaN   
2   1  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   0  U dun say so early hor... U c already then say...        NaN   
4   0  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [20]:
# mengubah nama kolom v1 menjadi labels dan v2 menjadi sms
data = data.rename(columns={'v1': 'labels', 'v2': 'sms'})

# memisahkan data independent dan dependent
X = data['sms']
y = data['labels']

# normalisasi data
vectorizer = TfidfVectorizer(stop_words='english')
X = vectorizer.fit_transform(X)

# split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# train model
model = MultinomialNB()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[963   0]
 [ 28 124]]
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       963
           1       1.00      0.82      0.90       152

    accuracy                           0.97      1115
   macro avg       0.99      0.91      0.94      1115
weighted avg       0.98      0.97      0.97      1115


### Kesimpulan
Dari hasil evaluasi, model Multinomial Naive Bayes dengan fitur CountVectorizer menghasilkan akurasi yang lebih baik dibandingkan dengan fitur TfidfVectorizer. Dengan nilai akurasi sebesar 0.98, sedangkan TfidfVectorizer sebesar 0.97.